File: test_06/src/main/java/MyStack.java
import java.util.NoSuchElementException;
/**
 * This class is my realisation of Stack.
 */
public class MyStack<E>  {
  private Node<E> top ;
  private int size ;

  /**
   * This is constructor.
   */
  public MyStack() {
    top = null;
    size = 0;
  }

  /**
   * Function to check if stack is empty
   * @return true if function is empty, false otherwise.
   */
  public boolean isEmpty() {
    return top == null;
  }

  /**
   * Function to get the size of the stack
   * @return size of the stack.
   */
  public int getSize() {
    return size;
  }

  /**
   * Function to push an element to the stack
   * @param data is element for pushing to stack.
   */
  public void push(E data) {
    Node pushValue = new Node (data, null);
    if (top == null)
      top = pushValue;
    else {
      pushValue.setNext(top);
      top = pushValue;
    }
    size++ ;
  }

  /**
   * Function to pop an element from the stack
   * @return
   */
  public E pop() {
    if (isEmpty() )
      throw new NoSuchElementException("Underflow Exception");
    Node deletingValue = top;
    top = deletingValue.getNext();
    size-- ;
    return (E) deletingValue.getData();
  }

  /**
   * Function to check the top element of the stack
   * @return return object at the top of this stack without removing it from the stack.
   */
  public E peek() {
    if (isEmpty() )
      throw new NoSuchElementException("Underflow Exception") ;
    return top.getData();
  }

  /**
   * This function clear stack.
   */
  public void clear() {
    if (isEmpty())
      throw new NoSuchElementException("Underflow Exception");
    while (size != 0) {
      Node deletingValue = top;
      top = deletingValue.getNext();
      size--;
    }
  }
}

File: TaskAboutStack/MyStack.cs
using System;

namespace TaskAboutStack
{
  /// <summary>
  /// Representation of stack.
  /// </summary>
  /// <typeparam name="Type">Type of elements in the stack.</typeparam>
  public class MyStack<Type>
  {
    private Element<Type> head;
    private Element<Type> tail;
    public int count { get; private set; }

    /// <summary>
    /// Default constructor.
    /// </summary>
    public MyStack()
    {
      head = null;
      tail = null;
      count = 0;
    }

    /// <summary>
    /// Returns the element at the top of the stack without removing it.
    /// </summary>
    /// <returns>element at the top of the stack</returns>
    public Type Peek()
    {
      if (IsEmpty())
      {
        throw new InvalidOperationException("Stack is empty.");
      }
      return tail.value;
    }

    /// <summary>
    /// Inserts an element at the top of the stack.
    /// </summary>
    /// <param name="value">Value of the element.</param>
    public void Push(Type value)
    {
      Element<Type> newElement = new Element<Type>(value);
      if (head == null)
      {
        head = newElement;
        tail = newElement;
      }
      else
      {
        tail.next = newElement;
        tail = newElement;
      }
      count++;
    }

    /// <summary>
    /// Returns the element at the top of the stack with removing it.
    /// </summary>
    /// <returns>Element at the top of the stack</returns>
    public Type Pop()
    {
      if (IsEmpty())
      {
        throw new InvalidOperationException("Stack is empty.");
      }

      Element<Type> topOfOtheStack = tail;
      Element<Type> current = head;
      while (current != null)
      {
        if (current.next == tail)
        {
          current.next = null;
          tail = current;
        }
        current = current.next;
      }
      count--;
      return topOfOtheStack.value;
    }

    /// <summary>
    /// Clears the stack.
    /// </summary>
    public void Clear()
    {
      if (IsEmpty())
      {
        return;
      }
      Element<Type> current = tail;
      while (count > 0)
      {
        this.Pop();
      }
    }

    /// <summary>
    /// Checks if the stack is empty.
    /// </summary>
    /// <returns>True is the stack is empty;otherwise false.</returns>
    public bool IsEmpty()
    {
      return count == 0;
    }
  }
}

